Create a Complement type

mkitti opened the pull request:
This adds a `Complement{C,T,N}` type.

The complement is to be interpreted as a lazy `complement`.

<strike>Math on `Complement` affects its underlying `Colorant`.</strike>

mkitti commented:
I will probably move this into its file that is included near the top of the file.

timholy commented on src/ColorVectorSpace.jl:
```diff
@@ -237,6 +238,91 @@ an array: `complement.(x)`
+* `Gray(0)` represents black. `Complement(Gray(0))` represents white.
+* `RGB(1,0,0)` represents red. `Complement(RGB(1,0,0))` represents cyan.
+
+Mathematical operations on `Complement` will affect the underlying
+`Colorant` rather than changing the value of the complement. That is
```
I'm worried about this behavior---generally, we value the interpretation above the representation. What should `Complement(0.1) + 0.1` return? I'd say 1.0 and not 0.2. `Complement(0.0)` is white, but with this logic, `Complement(0.0) == 1.0` should return `false`.

mkitti replied:
Perhaps `Complement(0.1) + 0.1` should be `Complement(0.0)` then. That is `Complement(x) + y` is `Complement(x-y)`.

timholy reviewed (commented):
I like introducing `Complement` but I worry the interpretation here isn't consistent. Perhaps you can explain more about why you think it should work this way?

mkitti commented:
I'm tempted to drop the math part from line 315 down

tlnagy commented:
Bump, this would be nice to have merged for https://github.com/tlnagy/TiffImages.jl/pull/134

mkitti commented on src/ColorVectorSpace.jl:
```diff
@@ -237,6 +238,62 @@ an array: `complement.(x)`
+`convert` to `C` at which point `complement` will be applied. Like `complement`
+the alpha channel is not modified.
+
+The main interpretation of this type is to invert the *interpretation* of the
```
```suggestion
The main application of this type is to invert the *interpretation* of the
```

mkitti commented:
@tlnagy what are the minimum features that you need?

tlnagy commented:
I think it would be nice for `Complement`ed values to be realized anytime you look at them. So it would be lazy until you accessed the values (like for display)

```julia
julia> res = Gray{N0f8}.(Matrix(I, 5, 5));

julia> Complement.(res) # doesn't work

TypeError: in typeassert, expected Complement{N0f8, 1, Gray{N0f8}}, got a value of type Gray{N0f8}

Stacktrace:
 [1] setindex!
   @ ./array.jl:971 [inlined]
 [2] setindex!
   @ ./multidimensional.jl:670 [inlined]
 [3] macro expansion
   @ ./broadcast.jl:974 [inlined]
 [4] macro expansion
   @ ./simdloop.jl:77 [inlined]
 [5] copyto!
   @ ./broadcast.jl:973 [inlined]
 [6] copyto!
   @ ./broadcast.jl:926 [inlined]
 [7] copy
   @ ./broadcast.jl:898 [inlined]
 [8] materialize(bc::Base.Broadcast.Broadcasted{Base.Broadcast.DefaultArrayStyle{2}, Nothing, Type{Complement}, Tuple{Matrix{Gray{N0f8}}}})
   @ Base.Broadcast ./broadcast.jl:873
 [9] top-level scope
   @ In[12]:1

julia> convert.(Gray{N0f8}, Complement.(res)) # does work, but clunky
```

mkitti commented:
Wait... why doesn't `Complement.(res)` work?

```julia
julia> using ColorVectorSpace, Colors, FixedPointNumbers, LinearAlgebra

julia> res = Gray{N0f8}.(Matrix(I, 5, 5));

julia> Complement.(res)
5×5 Array{Complement{N0f8},2} with eltype Complement{N0f8, 1, Gray{N0f8}}:
 Complement(Gray{N0f8}(1.0))  Complement(Gray{N0f8}(0.0))  …  Complement(Gray{N0f8}(0.0))  Complement(Gray{N0f8}(0.0))
 Complement(Gray{N0f8}(0.0))  Complement(Gray{N0f8}(1.0))     Complement(Gray{N0f8}(0.0))  Complement(Gray{N0f8}(0.0))
 Complement(Gray{N0f8}(0.0))  Complement(Gray{N0f8}(0.0))     Complement(Gray{N0f8}(0.0))  Complement(Gray{N0f8}(0.0))
 Complement(Gray{N0f8}(0.0))  Complement(Gray{N0f8}(0.0))     Complement(Gray{N0f8}(1.0))  Complement(Gray{N0f8}(0.0))
 Complement(Gray{N0f8}(0.0))  Complement(Gray{N0f8}(0.0))     Complement(Gray{N0f8}(0.0))  Complement(Gray{N0f8}(1.0))
```

I see that it should display, but it should "work".

mkitti commented:
I think the problem is that `Complement` is a `Colorant` rather than a `Color`. Perhaps it should be a `Color`.

mkitti commented:
I made `Complement <: Color`, so now it renders in VS Code. I'm not sure if that is the best design at the moment. Maybe I should write particular display for `Complement`?

tlnagy commented:
Would it be possible to write a display for Complement that just punts to the correct display based on the wrapped type? i.e. `Complement{T, N, C <:  Colorant{T,N}}` realizes the transform and then punts to the display for `Colorant{T, N}` etc

mkitti commented:
I believe so.

My understanding is that this is implemented in Colors.jl:

https://github.com/JuliaGraphics/Colors.jl/blob/master/src%2Fdisplay.jl

It looks like we just need to overload `show` for `Complement`. I think the implementation could be simple in that we would do the conversion to `C`, the underlying `Colorant`, and then invoke its `show` method.

mkitti commented:
I have an implementation of display now.

```julia
julia> using ColorVectorSpace, Colors, FixedPointNumbers, LinearAlgebra

julia> res = rand(RGB{Float64}, 10, 10)

julia> c_res = reinterpret(Complement, res) # less allocations

juila> using TestImages

julia> cameraman = testimage("cameraman.tif")

juila> c_cameraman = reinterpret(Complement, cameraman) # no allocation complement

julia> ca_cameraman = ComplementArray(c_cameraman) # no allocation, now of element type Gray{N0f8}

julia> [cameraman ca_cameraman]
```

mkitti commented:
I'm getting tempted to implement this in its own package. It seems slightly out of place here and we might be able to iterate on it faster if it were independent.

tlnagy commented:
Looks great! There's a slight issue though with display for larger images:

```julia
using ColorVectorSpace
using Colors
using FixedPointNumbers
using ImageShow
using LinearAlgebra

res = Gray{N0f8}.(Matrix(I, 500, 500)) # no warning
#snip

@time reinterpret(Complement, res) # works but with warning
#snip
┌ Warning: Output swatches are reduced due to the large size (500×500).
│ Load the ImageShow package for large images.
└ @ Colors ~/.julia/packages/Colors/mIuXl/src/display.jl:15
```
This happens *despite* ImageShow being loaded and this warning doesn't happen for non-Complemented images.

mkitti commented:
Should we push to get this merged now or should I break out into a package to iterate on this independently?

tlnagy commented:
Given that `complement` lives here, I think it makes sense to have the lazy version here as well. But I'm willing to defer to other folks who know the `Colors` ecosystem better. Would be nice to have merged either way.

timholy commented:
Sorry, I'll try to get this a review by next Monday. I got started and remembered getting worried about something fundamental, but that was long enough ago that I've forgotten.

mkitti commented:
There's been considerable changes since you last took a look, Tim.
1. `Complement` subtypes `Color` instead of `Colorant`
2. I changed the order of parameters from `{T,N,C}` to `{C,T,N}`
3. I dropped arithmetic on `Complement`
4. Accessing the components of `Complement` returns the `complement` of the parent `Colorant`.

mkitti commented:
Implementation of `Base.reinterpret(::Type{Complement}, array::AbstractArray{C}) where {T, N, C <: Colorant{T,N}}` may be an abuse because `Complement` by itself is not fully specified with parameters. However, it is inferrable that we mean `Complement{C,T,N}`

timholy commented on src/ColorVectorSpace.jl:
```diff
@@ -237,6 +238,112 @@ an array: `complement.(x)`
+Complement{C, <: Any, N}(args::Vararg{T,N}) where {C <: Colorant,N,T} = Complement(C(args...))
+
+@inline Base.parent(c::Complement) = getfield(c, :parent)
+Base.getproperty(c::Complement, s::Symbol) = getproperty(parent(c), s)
```
```suggestion
function Base.getproperty(c::Complement, s::Symbol)
    s === :alpha && return getproperty(parent(c), :alpha)
    return complement(getproperty(parent(c), s))
end
```

timholy commented on src/ColorVectorSpace.jl:
```diff
@@ -237,6 +238,112 @@ an array: `complement.(x)`
+`Complement{T}` element type. This allows for a `AbstractArray{Complement{T}}`
+to be wrapped into an `AbstractArray{T}` without additional allocations.
+"""
+struct ComplementArray{T,N,A} <: AbstractArray{T,N}
```
```suggestion
struct ComplementArray{T,N,A<:AbstractArray{T,N}} <: AbstractArray{T,N}
```

timholy commented on src/ColorVectorSpace.jl:
```diff
@@ -237,6 +238,112 @@ an array: `complement.(x)`
+Base.parent(a::ComplementArray) = a.parent
+Base.size(a::ComplementArray) = size(a.parent)
+function Base.getindex(a::ComplementArray{T}, I::Vararg{Int,N})::T where {T,N}
+    getindex(a.parent, I...)
```
I'm confused, shouldn't this be `complement(getindex(a.parent, I...))`? Basically this works like `mappedarray(complement, A)`, right?

Which begs the question, do we actually need `ComplementArray`? But perhaps I misunderstand the intention.

timholy commented on src/ColorVectorSpace.jl:
```diff
@@ -237,6 +238,112 @@ an array: `complement.(x)`
+    getindex(a.parent, I...)
+end
+function Base.setindex!(a::ComplementArray{T}, v, I::Vararg{Int, N}) where {T,N}
+    setindex!(a.parent, convert(eltype(a.parent), v))
```
The `convert` will happen automatically, no need to do it manually.

timholy commented on src/ColorVectorSpace.jl:
```diff
@@ -237,6 +238,112 @@ an array: `complement.(x)`
+function Base.setindex!(a::ComplementArray{T}, v, I::Vararg{Int, N}) where {T,N}
+    setindex!(a.parent, convert(eltype(a.parent), v))
+end
+Base.IndexStyle(a::ComplementArray) = IndexStyle(a.parent)
```
If you do this you should also implement `getindex` for a single `Int`.

timholy commented on src/ColorVectorSpace.jl:
```diff
@@ -237,6 +238,112 @@ an array: `complement.(x)`
+Base.IndexStyle(a::ComplementArray) = IndexStyle(a.parent)
+
+function Base.show(io::IO, mime::MIME"image/svg+xml", c::Complement{C}) where C
+    show(io, mime, C(c))
```
What happens without this method? I would think that if you have all the abstractions correct, this Just Works without needing a new method. Similar thoughts about the other methods below.

timholy reviewed (commented):
I'm on board with the general idea. Needs tests.

mkitti commented:
I agree. I just have to find some time to get back around to it. Feel free to beat me to it.

mkitti reviewed (commented):
Commit suggestions from Tim Holy

mkitti commented:
@timholy and @tlnagy I have written tests.

timholy reviewed (commented):
Thanks for adding the tests and other fixes! It's looking good, but I've left a few other notes about things that seem likely to need fixing.

In cases where you're unsure how something should behave, I tend to expect that `mappedarray(complement, a)` should be a reasonable guide for how `ComplementArray` should behave. IF, that is, I understand why it's here at all.

mkitti commented:
> Thanks for adding the tests and other fixes! It's looking good, but I've left a few other notes about things that seem likely to need fixing.
> 
> In cases where you're unsure how something should behave, I tend to expect that `mappedarray(complement, a)` should be a reasonable guide for how `ComplementArray` should behave. IF, that is, I understand why it's here at all.

* `mappedarray(complement, a)` is a good choice if you have an array of `C <: Colorant` and want an `AbstractArray{C}` that has an element type of `C`.
* `ComplementArray` is for when you have a `AbstractArray{Complement}` and want a `AbstractArray{C} where C <: Colorant`. Now that you mention `mappedarray`, it could probably just be a `mappedarray(parent, a)`.

kimikage commented:
My main concern is operation and support in the *downstream* packages.
```julia
julia> Complement(RGB(1,1,0)) isa AbstractRGB
false
```
It is not very pleasant to have to wrap `AbstractRGB` (or `AbstactGray`) with `Union`, in order to support `Complement`.
One ad hoc solution is to define `ComplementRGB{T} <: AbstractRGB{T}`, etc., respectively.😑

kimikage commented:
For reference, `AbstractGray` will be changed from an alias to an abstract type (sometime in the future). 
cf. https://github.com/JuliaGraphics/ColorTypes.jl/pull/252
Ironically, the change is to address this issue properly. (https://github.com/JuliaGraphics/ColorTypes.jl/issues/230)

timholy commented:
The other option is to convert `AbstractRGB` into a trait. That's feasible but a little harder to work with (given that traits do not have a standard convenient interface) than leaving it as an abstract type.

mkitti commented:
@tlnagy The outstanding question here is if `mappedarray(complement, a)` would satisfy the question you put forth in https://github.com/tlnagy/TiffImages.jl/pull/134 . Is that consistent with "Colors.jl promise to faithfully capture the interpretation of the raw data"?

To clearly indicate that I am not intending further action here until a downstream application arises, I am closing this pull request. I would be willing to reopen if a need arises.
